modules/ismp/pallet/pallet-asset-transfer/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+    }
+
+    pub fn dot_asset_id() -> H256 {
+        Params::<T>::get().unwrap_or(<T as Config>::Params::get()).dot_asset_id
+    }
+
+    /// Dispatch ismp request to token gateway on destination chain
+    pub fn dispatch_request(
+        multi_account: MultiAccount<T::AccountId>,
+        amount: <T::Assets as fungibles::Inspect<T::AccountId>>::Balance,
+    ) -> Result<(), Error<T>> {
+        let dispatcher = Dispatcher::<T>::default();
+
+        let mut to = [0u8; 32];
+        to[..20].copy_from_slice(&multi_account.evm_account.0);
+        let from: [u8; 32] = multi_account.substrate_account.clone().into();
+        let asset_id = Self::dot_asset_id().0.into();
+        let body = Body {
+            amount: {
+                let amount: u128 = amount.into();
+                let mut bytes = [0u8; 32];
+                U256::from(amount).to_big_endian(&mut bytes);
+                alloy_primitives::U256::from_be_bytes(bytes)
+            },
+            asset_id,
+            redeem: false,
+            from: from.into(),
+            to: to.into(),
+        };
+
+        let dispatch_post = DispatchPost {
+            dest: multi_account.dest_state_machine,
+            from: Self::token_gateway_address().0.to_vec(),
+            to: Self::token_gateway_address().0.to_vec(),
+            timeout_timestamp: multi_account.timeout,
+            data: alloy_rlp::encode(body),
+        };
+
+        dispatcher
+            .dispatch_request(
+                DispatchRequest::Post(dispatch_post),
+                multi_account.substrate_account.clone(),
+                Default::default(),
+            )
+            .map_err(|_| Error::<T>::DispatchPostError)?;
+
+        Self::deposit_event(Event::<T>::TransferInitiated {
+            from: multi_account.substrate_account,
+            to: multi_account.evm_account,
+            dest: multi_account.dest_state_machine,
+            amount,
+        });
+
+        Ok(())
+    }
+}
+
+#[derive(RlpDecodable, RlpEncodable, Debug, Clone)]
+pub struct Body {
+    // amount to be sent
+    pub amount: alloy_primitives::U256,
+    // The token identifier
+    pub asset_id: alloy_primitives::B256,
+    // flag to redeem the erc20 asset on the destination
+    pub redeem: bool,
+    // sender address
+    pub from: alloy_primitives::B256,
+    // recipient address
+    pub to: alloy_primitives::B256,
+}
+
+#[derive(Clone)]
+pub struct Module<T>(PhantomData<T>);
+
+impl<T: Config> Default for Module<T> {
+    fn default() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<T: Config> IsmpModule for Module<T>
+where
+    <T::Assets as fungibles::Inspect<T::AccountId>>::Balance: From<u128>,
+    u128: From<<T::Assets as fungibles::Inspect<T::AccountId>>::Balance>,
+    <T::Assets as fungibles::Inspect<T::AccountId>>::AssetId: From<MultiLocation>,
+    T::AccountId: Into<[u8; 32]> + From<[u8; 32]>,
+{
+    fn on_accept(&self, post: ismp::router::Post) -> Result<(), ismp::error::Error> {
```

Comment:
We should also charge protocol fees here

Reply:
I thought gateway will do that when initializing the transaction on the evm that's why I removed the charge there.

Reply:
> I thought gateway will do that when initializing the transaction on the evm that's why I removed the charge there.

It doesn't charge fees for `ERC6160` assets. It only charges when redeeming `ERC20` assets